selfdrive/boardd/boardd.cc

```diff
@@ -561,34 +588,55 @@ int main(int argc, char* argv[]) {
 
   while (!do_exit) {
     std::vector<std::thread> threads;
+    std::vector<Panda *> pandas;
+    Panda *peripheral_panda;
 
-    Panda *peripheral_panda = usb_connect(peripheral_panda_serial);
-    Panda *panda = (peripheral_panda_serial == panda_serial) ? peripheral_panda : usb_connect(panda_serial);
+    pandas = Panda::device_list();
+    for (const auto& panda : pandas){
+      panda_init(panda);
+    }
 
     // Send empty pandaState & peripheralState and try again
-    if (panda == nullptr || peripheral_panda == nullptr) {
+    if (pandas.size() == 0) {
       send_empty_panda_state(&pm);
       send_empty_peripheral_state(&pm);
       util::sleep_for(500);
       goto fail;
     }
 
+    // Sort to make sure the pandas are always in a deterministic order
+    std::sort(pandas.begin(), pandas.end(),
+      [](const Panda *a, const Panda *b) -> bool {
+        // Make sure the internal one is always first
+        if (a->is_internal && !b->is_internal) return true;
+        if (!a->is_internal && b->is_internal) return false;
+
+        // Sort by hardware type
+        if (a->hw_type != b->hw_type) {
+          return a->hw_type > b->hw_type;
+        }
+
+        // Last resort, sort by serial
+        return a->usb_serial.compare(b->usb_serial);
+      }
+    );
```

Comment:
This needs testing. Can never remember what returning true/false exactly does to the order of the sorted list